lib/client.ts

```diff
@@ -17,58 +24,156 @@ export class Client {
+   */
   public getAuthUri = async (opts: GetAuthUriOpts = {}): Promise<string> => {
     const query: {
       [key: string]: string | string[];
     } = {};
 
-    if (opts.codeChallenge) {
-      query.code_challenge = opts.codeChallenge;
-    }
+    if (opts.verifier) {
+      // Implemented according to https://tools.ietf.org/html/rfc7636#appendix-A
+      const challenge =
+        (btoa(String.fromCharCode.apply(null, sha256.array(opts.verifier))) || "")
+          .split("=")[0]
+          .replace("+", "-")
+          .replace("/", "_")
 
-    if (opts.codeChallengeMethod) {
-      query.code_challenge_method = opts.codeChallengeMethod;
+      query.code_challenge = challenge;
+      query.code_challenge_method = 'S256';
     }
 
-    const uri = await this.oauth2Client.code.getUri({
-      query,
-      state: this.state
-    });
-
-    return uri;
+    return this.oauth2Client.code.getUri({ query });
   };
 
+  /**
+   * This method must be called during the callback via `redirectUri`.
+   */
   public getToken = async (
     callbackUri: string,
     opts: GetTokenOpts = {}
   ): Promise<ClientOAuth2.Token> => {
     const options: {
-      state: string;
       body?: {
         code_verifier: string;
       };
-    } = {
-      state: this.state
-    };
+    } = {};
 
-    if (opts.codeVerifier) {
+    if (opts.verifier) {
       options.body = {
-        code_verifier: opts.codeVerifier
+        code_verifier: opts.verifier
       };
     }
 
-    const tokenData = await this.oauth2Client.code.getToken(
-      callbackUri,
-      options
+    const token = await this.oauth2Client.code.getToken(callbackUri, options);
+
+    this.token = token;
+
+    return token;
+  };
+
+  /**
+   * Use a previously created token for all upcoming requests.
+   */
+  public setToken = (
+    accessToken: string,
+    refreshToken?: string,
+    tokenType?: string
+  ): ClientOAuth2.Token => {
+    const data = {};
+    let token;
+
+    if (tokenType && refreshToken) {
+      token = this.oauth2Client.createToken(
+        accessToken,
+        refreshToken,
+        tokenType,
+        data
+      );
+    } else if (refreshToken) {
+      token = this.oauth2Client.createToken(accessToken, refreshToken, data);
+    } else {
+      token = this.oauth2Client.createToken(accessToken, data);
+    }
+
+    this.token = token;
+    return token;
+  };
+
+  /**
+   * Send a raw GraphQL request and return its response.
+   */
+  public rawQuery = async (
+    query: string,
+    variables?: Variables
+  ): Promise<Query> => {
+    if (!this.token) {
+      throw new Error("User unauthorized");
+    }
+
+    this.graphQLClient.setHeader(
+      "Authorization",
+      `Bearer ${this.token.accessToken}`
     );
-    return tokenData;
+
+    const { data } = await this.graphQLClient.rawRequest(query, variables);
+
+    return data;
+  };
+
+  public getAllTransactions = async (): Promise<Array<TransactionListItem>> => {
```

Comment:
What about moving this to a separate file into a method `getAll()` or `fetch()` and then include it here in the client? The user could call it with `await client.transactions.getAll()`

Reply:
sure, i had a plan to split it, this file cannot grow indefinitely :)